### PR TITLE
rewrite mysql

### PR DESCRIPTION
**Pull request by plally**

### CODE REVIEW

**Comment by sarahsturgeon** [comment]
Style or syntax errors were detected

**Comment by sarahsturgeon** [comment]
Style or syntax errors were detected

**Comment by sarahsturgeon** [comment]
Great job!

**Comment by sarahsturgeon** [comment]
Style or syntax errors were detected

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (plally) [inline lua/cfcfactions/core/sv/sv_mysql.lua:250]: haven't really done anything with user permissions because im not sure how those should be stored. 
Json? 
bigint (each bit being a Boolean permission) ?

Comment (ghost) [reply lua/cfcfactions/core/sv/sv_mysql.lua:250]: Permissions are stored as a table of strings per user. There are actually 2 tables of permissions, 1 is global cfc permissions that are special, the other is a specific faction's permissions 

I think JSON would be better for storing. See what Lego thinks first

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (ghost) [inline lua/cfcfactions/core/sv/sv_mysql.lua:39]: Colors in gmod are a table, which I used as a string. So this is going to be messy. Do you have anything for Table to Int for colors specifically. Its Color(R,G,B,A)

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (ghost) [inline lua/cfcfactions/core/sv/sv_mysql.lua]: Faction permissions is missing. A faction has a table of strings just for their specific permission. JSON preferred.

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (plally) [reply lua/cfcfactions/core/sv/sv_mysql.lua:39]:
```lua
function rgbToInt(r,g,b)
    return bit.lshift( r, 16 ) + bit.lshift( g, 8 ) + b
end
```

Comment (plally) [reply lua/cfcfactions/core/sv/sv_mysql.lua:39]: i could add some functions to convert colors to and from an int if that helps
is it safe to assume the alpha doesn't matter?

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (plally) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: what kind of permissions are in a factions permissions table?

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (ghost) [reply lua/cfcfactions/core/sv/sv_mysql.lua:39]: You would be correct. Color in gmod is a table, just let us pass it as a Color(table) instead of each rgb

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (ghost) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: https://github.com/brandonsturgeon/cfc_factions_3/blob/master/lua/cfcfactions/core/sv/sv_permsys.lua


Table will contain a string of whatever permission is required for user. {} = "CanShoot", "CanJump" example.

Comment (plally) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: but why does a faction need this table wouldn't that just be per user? is it like a list of possible permissions that someone in that faction can have?

Comment (ghost) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: Similar to how ulx, a user can have different permissions changed about them. A user can kick from a faction where another user cannot. Factions, can make their own little ranks with their own permission system. That is what phatso wanted long ago, might want to see what he thinks now. 2 factions can both contain a rank called "Admin" but they might have different permissions depending on the owner of the faction.

Comment (plally) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: we might want to separate this into its own table
```
faction
rank
permissions
```

Comment (ghost) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: I'd agree with that. That is also ontop of user permissions for factions. Maybe we could get away with storing them as ulx permissions instead. 

Or maybe simplify shit to just "Admin, Owner, and User" See what others including phatso thinks about this

Comment (plally) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: could store just the admin permissions, such as deleting any faction, as ulx/ulib permissions.

Comment (sarahsturgeon) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: It would be nice to let the factions choose their own rank names to a degree, so I like the way you have it here if you just separate them out like PCL suggested

Comment (ghost) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: Make a card for this on board then

Comment (plally) [reply lua/cfcfactions/core/sv/sv_mysql.lua]: so store user permissions such as IsDeveloper IsAdmin etc in the users sql table
store faction permissions such as CanInvite CanEdit etc in the ranks sql table

**Comment by sarahsturgeon** [comment]
Where's this at? I think we need to wait for V's big PR to get merged first before merging this

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (justinmuskopf) [inline lua/cfcfactions/core/sv/sv_mysql.lua:10]: I'm used to seeing `local myqloo = require( 'mysqloo' )`.
Is there any reason we're separating it out?

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (justinmuskopf) [inline lua/cfcfactions/core/sv/sv_mysql.lua:30]: `initilize` --> `initialize` 
Doing God's work, son.

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (justinmuskopf) [inline lua/cfcfactions/core/sv/sv_mysql.lua:137]: Remove debug print or give it a logging header (e.g. "[CFC Factions 3] Performing SQL query: {query}")

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (justinmuskopf) [inline lua/cfcfactions/core/sv/sv_mysql.lua:136]:
```suggestion
function sql_db:doQuery( queryString, successCallback, errorCallback )
```

File: lua/cfcfactions/core/sv/sv_mysql.lua
Comment (justinmuskopf) [inline lua/cfcfactions/core/sv/sv_mysql.lua:163]: Should we `nil` check these fields?

**Review by justinmuskopf — APPROVED** [review]

**Pull request merged**
